TDataFrame.hxx

```diff
@@ -856,182 +1070,172 @@ public:
+      // forget actions and "detach" the action result pointers marking them ready and forget them too
       fBookedActions.clear();
-      fBookedFilters.clear();
-      fBookedBranches.clear();
       for (auto aptr : fActionResultsPtrs) {
          aptr->SetReady();
       }
       fActionResultsPtrs.clear();
    }
 
    // build reader values for all actions, filters and branches
-   void BuildAllReaderValues(TTreeReader& r, unsigned slot) {
-      for(auto& ptr : fBookedActions)
-         ptr->BuildReaderValues(r, slot);
-      for(auto& ptr : fBookedFilters)
-         ptr->BuildReaderValues(r, slot);
-      for(auto& bookedBranch : fBookedBranches)
-         bookedBranch.second->BuildReaderValues(r, slot);
+   void BuildAllReaderValues(TTreeReader &r, unsigned int slot)
+   {
+      for (auto &ptr : fBookedActions) ptr->BuildReaderValues(r, slot);
+      for (auto &ptr : fBookedFilters) ptr->BuildReaderValues(r, slot);
+      for (auto &bookedBranch : fBookedBranches) bookedBranch.second->BuildReaderValues(r, slot);
    }
 
    // inform all actions filters and branches of the required number of slots
-   void CreateSlots(unsigned nSlots) {
-      for(auto& ptr : fBookedActions)
-         ptr->CreateSlots(nSlots);
-      for(auto& ptr : fBookedFilters)
-         ptr->CreateSlots(nSlots);
-      for(auto& bookedBranch : fBookedBranches)
-         bookedBranch.second->CreateSlots(nSlots);
+   void CreateSlots(unsigned int nSlots)
+   {
+      for (auto &ptr : fBookedActions) ptr->CreateSlots(nSlots);
+      for (auto &ptr : fBookedFilters) ptr->CreateSlots(nSlots);
+      for (auto &bookedBranch : fBookedBranches) bookedBranch.second->CreateSlots(nSlots);
    }
 
-   TDataFrame& GetDataFrame() const {
-      return fFirstData;
-   }
+   std::weak_ptr<Details::TDataFrameImpl> GetDataFrame() const { return fFirstData; }
 
-   const BranchList& GetDefaultBranches() const {
-      return fDefaultBranches;
-   }
+   const BranchList &GetDefaultBranches() const { return fDefaultBranches; }
 
-   const TDataFrameBranchBase& GetBookedBranch(const std::string& name) const {
+   const TDataFrameBranchBase &GetBookedBranch(const std::string &name) const
+   {
       return *fBookedBranches.find(name)->second.get();
    }
 
-   void* GetTmpBranchValue(const std::string& branch, unsigned slot, int entry) {
+   void *GetTmpBranchValue(const std::string &branch, unsigned int slot, int entry)
+   {
       return fBookedBranches.at(branch)->GetValue(slot, entry);
    }
 
-   TDirectory* GetDirectory() const {
-      return fDirPtr;
-   }
+   TDirectory *GetDirectory() const { return fDirPtr; }
 
-   std::string GetTreeName() const {
-      return fTreeName;
-   }
+   std::string GetTreeName() const { return fTreeName; }
 
-private:
-   void Book(ActionBasePtr actionPtr) {
-      fBookedActions.emplace_back(std::move(actionPtr));
-   }
+   void Book(Internal::ActionBasePtr_t actionPtr) { fBookedActions.emplace_back(std::move(actionPtr)); }
```

Comment:
these `std::move`s were useful when we used `unique_ptr`s instead of `shared_ptr`s...I think we can remove them now.